cardano-api/src/Cardano/Api/Eras.hs

```diff
@@ -365,6 +383,62 @@ instance IsShelleyBasedEra BabbageEra where
 instance IsShelleyBasedEra ConwayEra where
    shelleyBasedEra = ShelleyBasedEraConway
 
+data AnyShelleyBasedEra where
+  AnyShelleyBasedEra
+    :: IsShelleyBasedEra era  -- Provide class constraint
+    => ShelleyBasedEra era    -- and explicit value.
+    -> AnyShelleyBasedEra
+
+deriving instance Show AnyShelleyBasedEra
+
+instance Eq AnyShelleyBasedEra where
+    AnyShelleyBasedEra era == AnyShelleyBasedEra era' =
+      case testEquality era era' of
+        Nothing   -> False
+        Just Refl -> True -- since no constructors share types
+
+instance Bounded AnyShelleyBasedEra where
+   minBound = AnyShelleyBasedEra ShelleyBasedEraShelley
+   maxBound = AnyShelleyBasedEra ShelleyBasedEraConway
```

Comment:
Do we have tests in place checking if all the instances here are correct, so that we could catch an error when adding new Era?

Reply:
I'm unaware of such tests existing.  Is there a way to encode such tests that doesn't merely move the problem to the tests?

Reply:
I don't think you can do that just on the type-level. I think it's more than just moving of the problem, and it gives you some level of assurance at the test execution level. In my opinion it would be worth having instances (longer than a single line) checked in tests, but I'm leaving it up to you if you'd prefer to go with it.

Reply:
I can at least confirmed `maxBound` for `AnyCardanoEra` and `AnyShelleyBasedEra` match, so I've added those tests.

I added some additional property tests for both as well.